pkg/controller/spec_processor.go

```diff
@@ -4,69 +4,90 @@ import (
-	selfName         smith_v1.ResourceName
-	resources        map[smith_v1.ResourceName]*resourceInfo
-	allowedResources map[smith_v1.ResourceName]struct{}
-	examplesOnly     bool
+	variables map[smith_v1.ReferenceName]interface{}
 }
 
 // noExampleError occurs when we try to process the spec with examples rather
 // than resolving references, but at least one of references doesn't specify an example.
 type noExampleError struct {
-	selector string
+	referenceName smith_v1.ReferenceName
 }
 
 func (e *noExampleError) Error() string {
-	return fmt.Sprintf("no example value provided in selector %q", e.selector)
+	return fmt.Sprintf("no example value provided in reference %q", e.referenceName)
 }
 
-func NewSpec(selfName smith_v1.ResourceName, resources map[smith_v1.ResourceName]*resourceInfo, allowedResources []smith_v1.ResourceName) *SpecProcessor {
-	return &SpecProcessor{
-		selfName:         selfName,
-		resources:        resources,
-		allowedResources: convertResourceNamesToMap(allowedResources),
-		examplesOnly:     false,
+func NewSpec(resources map[smith_v1.ResourceName]*resourceInfo, references []smith_v1.Reference) (*SpecProcessor, error) {
+	variables, err := resolveAllReferences(references, func(reference smith_v1.Reference) (interface{}, error) {
+		return resolveReference(resources, reference)
+	})
+
+	if err != nil {
+		return nil, err
 	}
-}
 
-func NewExamplesSpec(selfName smith_v1.ResourceName, allowedResources []smith_v1.ResourceName) *SpecProcessor {
 	return &SpecProcessor{
-		selfName:         selfName,
-		allowedResources: convertResourceNamesToMap(allowedResources),
-		examplesOnly:     true,
+		variables: variables,
+	}, nil
+}
+
+func NewExamplesSpec(references []smith_v1.Reference) (*SpecProcessor, error) {
+	variables, err := resolveAllReferences(references, func(reference smith_v1.Reference) (interface{}, error) {
+		if reference.Example == nil {
+			return nil, errors.WithStack(&noExampleError{referenceName: reference.Name})
+		}
+		return reference.Example, nil
+	})
+
+	if err != nil {
+		return nil, err
 	}
+
+	return &SpecProcessor{
+		variables: variables,
+	}, nil
 }
 
-func convertResourceNamesToMap(resources []smith_v1.ResourceName) map[smith_v1.ResourceName]struct{} {
-	ar := make(map[smith_v1.ResourceName]struct{}, len(resources))
-	for _, allowedResource := range resources {
-		ar[allowedResource] = struct{}{}
+func resolveAllReferences(
+	references []smith_v1.Reference,
+	resolveReference func(reference smith_v1.Reference) (interface{}, error),
+) (map[smith_v1.ReferenceName]interface{}, error) {
+
+	ar := make(map[smith_v1.ReferenceName]interface{}, len(references))
+	var finalerr error
+	for _, reference := range references {
+		// Don't 'resolve' nameless references - they're just being
+		// used to cause dependencies.
+		if reference.Name == "" {
+			continue
+		}
+
+		resolvedRef, err := resolveReference(reference)
+		if err != nil {
+			finalerr = multierr.Append(finalerr, err)
```

Comment:
Is this by design the there is no `continue` here to avoid adding invalid resolved ref into the map?